keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1452 @@
+  `Always` will be terminated when the pod terminates and has no influence over
+  the pod lifecycle. Also, for Pods, the default `restartPolicy` is
+  [documented](https://github.com/kubernetes/kubernetes/blob/280473ebc4e45f9001f5f9789c318ff7329bc5f0/staging/src/k8s.io/api/core/v1/types.go#L2753-L2764)
+  as `Always` but for `initContainers` it will default to `OnFailure`. We
+  believe this can be addressed in the documentation of the existing enum
+  fields.
+
+When in future we may support `Always` on regular containers, there will be
+interesting case of `Always` having a meaning of non-blocking container for the
+Pod with `restartPolicy == Never` or `OnFailure`. This is easy to explain - Pod
+lifecycle is controlled by containers with the matching restartPolicy. But it
+may be slightly confusing and needs to be carefully reviewed at the time this
+feature will be considered.
+
+  
+#### Alternatives
+
+- Introduce a new collection `infrastructureContainers` that replaces
+  semantically `initContainers` and deprecate the `initContainers`. This
+  collection will allow both - Init containers and sidecar containers. Decision
+  on this alternative can be postponed to after alpha implementation.
+- Introduce a new enum value
+  `UntilPodTermination`/`UntilPodShutdown`/`WithPod`/`Ambient` for
+  `restartPolicy` on containers. This alternative was rejected because the
+  `Always` semantic is quite easy to explain and potential confusion seems to be
+  minimal. Furthermore, if a new enum value is added, the question would be
+  whether we want to keep the `Always` value as an option and what it's semantic
+  will be. It is likely that semantic of `Always` will match almost exactly
+  the semantic of this new value.
+- Introduce a new field: `ambient: true` on all containers. This property will
+  make containers be restarted all the time, and will not block the pod
+  completion.
+  - Pros: detaching sidecar proposal from the per-container `restartPolicy`
+    proposal
+  - Cons: this field is a new concept that will be introduced for the same
+    property that is typically controlled by `restartPolicyß`.
+
+
+### User Stories (Optional)
+
+User stories are covered in Motivation section.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The following is the list of hypothetical scenarios how users may decide to
+abuse the feature or use it in a non-designed way. This is an exercise to
+understand potential side effects from implementing the feature. We are looking
+for side effects like abuse of the functionality to implement something
+error-prone that sidecars were not designed for. Or causing issues like noisy
+neighbors, etc.
```

Comment:
Is there a scenario here for ordering problems when attempting to inject two sidecars?  I haven't carefully thought through the mutating webhook semantics, but I seem to recall that the mutations are collected in parallel; I'm not sure how two injectors might agree on an order of container injection.

Reply:
Added scenario 1.a:

> ### Scenario 1.a
> 
>User decides to run regular container as a sidecar to simply guarantee the
> startup sequence of the containers. Users can already implement the startup sequencing by using blocking nature > of `postStart` hook. So using the sidecar
> containers instead is not adding any benefits.
>
> As for injecting sidecars into Pods with regular containers run as sidecars - 
> webhooks will likely inject sidecar containers to be first, so the risk is minimal here as well.

I cannot formulate to myself the risk related to different ordering of sidecars injection by the webhook

Reply:
I suppose the risk is no different than trying to inject `initContainers` without sidecars.